File: app/src/main/java/com/pluscubed/velociraptor/settings/ToleranceDialogFragment.java
package com.pluscubed.velociraptor.settings;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.ToggleButton;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.pluscubed.velociraptor.R;
import com.pluscubed.velociraptor.utils.PrefUtils;
import com.pluscubed.velociraptor.utils.Utils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ToleranceDialogFragment extends DialogFragment {

    @BindView(R.id.text_constant_unit)
    TextView constantUnitText;
    @BindView(R.id.edittext_constant)
    EditText constantEditText;
    @BindView(R.id.seekbar_constant)
    SeekBar constantSeekbar;

    @BindView(R.id.text_percent)
    TextView percentText;
    @BindView(R.id.edittext_percent)
    EditText percentEditText;
    @BindView(R.id.seekbar_percent)
    SeekBar percentSeekbar;

    @BindView(R.id.button_and)
    ToggleButton andButton;
    @BindView(R.id.button_or)
    ToggleButton orButton;

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        @SuppressLint("InflateParams")
        View dialog = getActivity().getLayoutInflater().inflate(R.layout.dialog_tolerance, null, false);
        ButterKnife.bind(this, dialog);

        constantUnitText.setText(Utils.getUnitText(getActivity()));
        constantEditText.setText(String.valueOf(PrefUtils.getSpeedingConstant(getActivity())));
        constantEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                try {
                    int constant = Integer.parseInt(s.toString());
                    constantSeekbar.setProgress(constant + 25);
                } catch (NumberFormatException e) {
                    constantSeekbar.setProgress(25);
                }
            }
        });
        constantSeekbar.setProgress(PrefUtils.getSpeedingConstant(getActivity()) + 25);
        constantSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    constantEditText.setText(String.valueOf(progress - 25));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        percentText.setText(getString(R.string.percent, ""));
        percentEditText.setText(String.valueOf(PrefUtils.getSpeedingPercent(getActivity())));
        percentEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                try {
                    int constant = Integer.parseInt(s.toString());
                    percentSeekbar.setProgress(constant + 25);
                } catch (NumberFormatException e) {
                    percentSeekbar.setProgress(25);
                }
            }
        });
        percentSeekbar.setProgress(PrefUtils.getSpeedingPercent(getActivity()) + 25);
        percentSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    percentEditText.setText(String.valueOf(progress - 25));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        andButton.setChecked(PrefUtils.getToleranceMode(getActivity()));
        orButton.setChecked(!PrefUtils.getToleranceMode(getActivity()));
        andButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                andButton.setChecked(true);
                orButton.setChecked(false);
            }
        });
        orButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                orButton.setChecked(true);
                andButton.setChecked(false);
            }
        });

        return new MaterialDialog.Builder(getActivity())
                .customView(dialog, true)
                .title(R.string.speeding_amount)
                .negativeText(android.R.string.cancel)
                .positiveText(android.R.string.ok)
                .onPositive(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        try {
                            PrefUtils.setSpeedingConstant(getActivity(), Integer.parseInt(constantEditText.getText().toString()));
                            PrefUtils.setSpeedingPercent(getActivity(), Integer.parseInt(percentEditText.getText().toString()));
                        } catch (NumberFormatException ignored) {
                        }
                        PrefUtils.setToleranceMode(getActivity(), andButton.isChecked());
                    }
                }).build();
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);

        Utils.updateFloatingServicePrefs(getActivity());
    }
}

File: app/src/main/java/com/pluscubed/velociraptor/settings/ToleranceDialogFragment.kt
package com.pluscubed.velociraptor.settings

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import android.widget.SeekBar
import android.widget.TextView
import android.widget.ToggleButton
import androidx.fragment.app.DialogFragment
import butterknife.BindView
import butterknife.ButterKnife
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.pluscubed.velociraptor.R
import com.pluscubed.velociraptor.utils.PrefUtils
import com.pluscubed.velociraptor.utils.Utils

class ToleranceDialogFragment : DialogFragment() {

    @BindView(R.id.text_constant_unit)
    lateinit var constantUnitText: TextView
    @BindView(R.id.edittext_constant)
    lateinit var constantEditText: EditText
    @BindView(R.id.seekbar_constant)
    lateinit var constantSeekbar: SeekBar

    @BindView(R.id.text_percent)
    lateinit var percentText: TextView
    @BindView(R.id.edittext_percent)
    lateinit var percentEditText: EditText
    @BindView(R.id.seekbar_percent)
    lateinit var percentSeekbar: SeekBar

    @BindView(R.id.button_and)
    lateinit var andButton: ToggleButton
    @BindView(R.id.button_or)
    lateinit var orButton: ToggleButton

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        @SuppressLint("InflateParams")
        val dialog = activity!!.layoutInflater.inflate(R.layout.dialog_tolerance, null, false)
        ButterKnife.bind(this, dialog)

        constantUnitText.text = Utils.getUnitText(activity!!)
        constantEditText.setText(PrefUtils.getSpeedingConstant(activity).toString())
        constantEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                try {
                    val constant = Integer.parseInt(s.toString())
                    constantSeekbar.progress = constant + 25
                } catch (e: NumberFormatException) {
                    constantSeekbar.progress = 25
                }

            }
        })
        constantSeekbar.progress = PrefUtils.getSpeedingConstant(activity) + 25
        constantSeekbar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    constantEditText.setText((progress - 25).toString())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        percentText.text = getString(R.string.percent, "")
        percentEditText.setText(PrefUtils.getSpeedingPercent(activity).toString())
        percentEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                try {
                    val constant = Integer.parseInt(s.toString())
                    percentSeekbar.progress = constant + 25
                } catch (e: NumberFormatException) {
                    percentSeekbar.progress = 25
                }

            }
        })
        percentSeekbar.progress = PrefUtils.getSpeedingPercent(activity) + 25
        percentSeekbar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    percentEditText.setText((progress - 25).toString())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        andButton.isChecked = PrefUtils.getToleranceMode(activity)
        orButton.isChecked = !PrefUtils.getToleranceMode(activity)
        andButton.setOnClickListener {
            andButton.isChecked = true
            orButton.isChecked = false
        }
        orButton.setOnClickListener {
            orButton.isChecked = true
            andButton.isChecked = false
        }

        return MaterialDialog(activity!!)
                .customView(view = dialog, scrollable = true)
                .title(R.string.speeding_amount)
                .negativeButton(android.R.string.cancel)
                .positiveButton(android.R.string.ok) {
                    try {
                        PrefUtils.setSpeedingConstant(
                                activity,
                                Integer.parseInt(constantEditText.text.toString())
                        )
                        PrefUtils.setSpeedingPercent(
                                activity,
                                Integer.parseInt(percentEditText.text.toString())
                        )
                    } catch (ignored: NumberFormatException) {
                    }

                    PrefUtils.setToleranceMode(activity, andButton.isChecked)
                }
    }

    override fun onDismiss(dialog: DialogInterface?) {
        super.onDismiss(dialog)

        Utils.updateFloatingServicePrefs(activity)
    }
}